analyzers/src/SonarAnalyzer.Common/Rules/Hotspots/PubliclyWritableDirectoriesBase.cs

```diff
@@ -0,0 +1,110 @@
+        private const string ForwardSlashThreePartTemplate = @"^\/{0}\/{1}\/{2}(\/|$)";
+        private const string Temp = @"^([a-z]:[\\\/]?|[\\\/][\\\/][^\\\/]+[\\\/]|[\\\/])(windows[\\\/])?te?mp([\/\\]|$)";
+        private readonly Regex userProfile = new Regex(@"^%USERPROFILE%[\\\/]AppData[\\\/]Local[\\\/]Temp([\\\/]|$)", WindowsAndUnixOptions);
+        private readonly string varTmp = string.Format(VariableSlashTwoPartTemplate, "var", "tmp");
+        private readonly string usrTmp = string.Format(VariableSlashTwoPartTemplate, "usr", "tmp");
+        private readonly string devShm = string.Format(VariableSlashTwoPartTemplate, "dev", "shm");
+        private readonly string devMqueue = string.Format(ForwardSlashTwoPartTemplate, "dev", "mqueue");
+        private readonly string runLock = string.Format(ForwardSlashTwoPartTemplate, "run", "lock");
+        private readonly string varRunLock = string.Format(ForwardSlashThreePartTemplate, "var", "run", "lock");
+        private readonly string libraryCaches = string.Format(ForwardSlashTwoPartTemplate, "library", "caches");
+        private readonly string usersShared = string.Format(ForwardSlashTwoPartTemplate, "users", "shared");
+        private readonly string privateTmp = string.Format(ForwardSlashTwoPartTemplate, "private", "tmp");
+        private readonly string privateVarTmp = string.Format(ForwardSlashThreePartTemplate, "private", "var", "tmp");
+        private readonly DiagnosticDescriptor rule;
+        private readonly Regex linuxPubliclyWritabelDirs;
+        private readonly Regex windowsAndMacPubliclyWritabelDirs;
+        private readonly Regex environmentVariables = new Regex($@"^%(temp|tmp|tmpdir)%([\\\/]|$)", WindowsAndUnixOptions);
+
+        protected abstract ILanguageFacade<TSyntaxKind> Language { get; }
+
+        private protected abstract bool IsGetTempPathAssignment(TInvocationExpression invocationExpression, KnownType type, string methodName, SemanticModel semanticModel);
+        private protected abstract bool IsInsecureEnvironmentVariableRetrieval(TInvocationExpression invocation, KnownType type, string methodName, SemanticModel semanticModel);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+
+        protected string[] InsecureEnvironmentVariables { get; } = { "tmp", "temp", "tmpdir" };
+
+        protected PubliclyWritableDirectoriesBase(IAnalyzerConfiguration configuration, ResourceManager rspecResources) : base(configuration)
+        {
+            rule = DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, rspecResources).WithNotConfigurable();
+            linuxPubliclyWritabelDirs = new Regex($"({devMqueue}|{runLock}|{varRunLock})", RegexOptions.Compiled);
+            windowsAndMacPubliclyWritabelDirs = new Regex($"({varTmp}|{usrTmp}|{devShm}|{libraryCaches}|{usersShared}|{privateTmp}|{privateVarTmp}|{Temp})", WindowsAndUnixOptions);
```

Comment:
Let's split this one to `windowsDirs` and `macDirs` to keep the logic separated per domain. It will allow remove `Temp` and initialize it directly in it's standalone regex.